ludwig/utils/hyperopt_utils.py

```diff
@@ -0,0 +1,333 @@
+#! /usr/bin/env python
+# coding=utf-8
+# Copyright (c) 2019 Uber Technologies, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+import logging
+from abc import ABC, abstractmethod
+from typing import Any, Dict
+
+from ludwig.constants import EXECUTOR, STRATEGY, MINIMIZE
+from ludwig.utils.defaults import default_random_seed
+from ludwig.utils.misc import get_from_registry, set_default_values, set_default_value
+
+logger = logging.getLogger(__name__)
+
+
+class HyperoptStrategy((ABC)):
+    def __init__(self, goal: str, parameters: Dict[str, Any]) -> None:
+        self.goal = goal  # useful for bayesian stratiegy
+        self.parameters = parameters
+
+    @abstractmethod
+    def sample(self) -> Dict[str, Any]:
+        # Yields a set of parameters names and their values.
+        # Define `build_hyperopt_strategy` which would take paramters as inputs
+        pass
+
+    @abstractmethod
```

Comment:
Hi @w4nderlust, Would it be necessary for the `update` method to be an `abstract` method? As stateless strategies like Grid and Random won't need it.

Reply:
As there will be statefull strategies, let's keep it, so that the executor can always call it *otherwise there will need to be a check in the excutor like `if hasattr(strategy, update)` which is not super nice. Strategies that make no use of it will just not implement it / implement it with `pass`.